### PR TITLE
Add a new key error type AssertionFailed

### PR DESCRIPTION
**Pull request by MyonKeminta**
This might be part of https://github.com/pingcap/tidb/issues/26833

### CODE REVIEW

**Review by disksing — APPROVED** [review]

**Comment by disksing** [comment]
Please take a look @ekexium @sticnarf

**Comment by sticnarf** [comment]
LGTM while I think we can keep it until the related PR is ready to merge.

File: proto/kvrpcpb.proto [hunk 1]
```diff
@@ -758,6 +759,14 @@ message CommitTsTooLarge {
     uint64 commit_ts = 1; // The calculated commit TS.
 }
 
+message AssertionFailed {
+    uint64 start_ts = 1;
+    bytes key = 2;
+    Assertion assertion = 3;
+    uint64 existed_start_ts = 4;
+    uint64 existed_commit_ts = 5;
```
Comment (ekexium) [inline proto/kvrpcpb.proto:767]:
```suggestion
    uint64 existing_start_ts = 4;
    uint64 existing_commit_ts = 5;
```

File: proto/kvrpcpb.proto [hunk 2]
```diff
@@ -721,6 +721,7 @@ message KeyError {
     CommitTsExpired commit_ts_expired = 7; // Commit ts is earlier than min commit ts of a transaction.
     TxnNotFound txn_not_found = 8; // Txn not found when checking txn status.
     CommitTsTooLarge commit_ts_too_large = 9; // Calculated commit TS exceeds the limit given by the user.
+    AssertionFailed assertion_failed = 10; // Assertion of a `Mutation` is evaluated to fail.
```
Comment (ekexium) [inline proto/kvrpcpb.proto:724]:
```suggestion
    AssertionFailed assertion_failed = 10; // Assertion of a `Mutation` is evaluated as a failure.
```

**Review by ekexium — APPROVED** [review]
LGTM

**Comment by disksing** [comment]
@MyonKeminta want to address comments?

**Comment by MyonKeminta** [comment]
> @MyonKeminta want to address comments?

Sorry I wasn't working on this for some time. I'll do it now

**Pull request merged**